GrayBMP.cs

```diff
@@ -169,3 +184,26 @@ void Fatal (string message)
    #endregion
 }
 #endregion
+
+#region Struct Point2D
+
+record struct Point2D (double X, double Y) {
+   /// <summary> Returns the projected point at given distance at given angle</summary>
+   public static Point2D Project (Point2D p, double d, double theta) => new (p.X + (d * Cos (theta)), p.Y + (d * Sin (theta)));
+   /// <summary> Returns vertices of polygon which encloses the given two points</summary>
+   public static Point2D[] GetPolygonVertices (Point2D startPt, Point2D endPt, double offset) {
+      double angle = Atan2 (endPt.Y - startPt.Y, endPt.X - startPt.X);
+      var pts = new Point2D[8];
+      int count = 0;
+      double startAngle = PI / 2,
+             endAngle = 3 * PI / 2,
+             increment = PI / 3;
+      for (double i = startAngle; i <= endAngle; i += increment) {
+         pts[count++] = Project (startPt, offset, angle + i);
+         pts[count++] = Project (endPt, -offset, angle + i);
+      }
+      return new Point2D[] { pts[0], pts[2], pts[4], pts[6], pts[1], pts[3], pts[5], pts[7], pts[0] };

```

Comment:
Try to avoid creating two arrays here on each call. Perhaps use two loops to generate the two end-caps separately instead of trying to do both in one loop? Even better, create an array in the calling function and pass it to GetPolygonVertices, so it can keep reusing the same array (rather than creating it here each time).